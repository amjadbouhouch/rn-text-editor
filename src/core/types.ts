import type {
  Mark as ProseMirrorMark,
  Node as ProseMirrorNode,
} from 'prosemirror-model';
import type { EditorState, Transaction } from 'prosemirror-state';
import type {
  Commands,
  ExtensionConfig,
  MarkConfig,
  NodeConfig,
} from 'rn-text-editor';
import type { Editor } from '.';
import type { Extension } from './Extension';
import type { Mark } from './Mark';
import type { Node } from './Node';

export type TextContentType = 'text' | 'paragraph';
type TextContentMark = {
  type?: 'italic' | 'bold' | 'link' | 'strike' | 'textStyle' | 'highlight';
  text?: string;
  attrs?: {
    href: string;
    color: string;
    label: string;
  };
};
type User = {
  id: string;
  displayName: string;
};

export type TextContent = {
  type: TextContentType;
  id: string;
  text?: string;
  marks?: TextContentMark[];
  content?: TextContent[];
  attrs?: {
    level: number;
    user?: User;
    start?: number;
    label?: string;
  };
  num?: number;
};

export type JSONContent = {
  type?: string;
  attrs?: Record<string, any>;
  content?: JSONContent[];
  marks?: {
    type: string;
    attrs?: Record<string, any>;
    [key: string]: any;
  }[];
  text?: string;
  [key: string]: any;
};
export type HTMLContent = string;

export type Content = HTMLContent | JSONContent | JSONContent[] | null;
export type FocusPosition = 'start' | 'end' | 'all' | number | boolean | null;
export type Range = {
  from: number;
  to: number;
};

export interface EditorEvents {
  beforeCreate: { editor: Editor };
  create: { editor: Editor };
  update: { editor: Editor; transaction: Transaction };
  selectionUpdate: { editor: Editor; transaction: Transaction };
  transaction: { editor: Editor; transaction: Transaction };
  focus: { editor: Editor; event: FocusEvent; transaction: Transaction };
  blur: { editor: Editor; event: FocusEvent; transaction: Transaction };
  destroy: void;
}
export type Diff<T extends keyof any, U extends keyof any> = ({
  [P in T]: P;
} & {
  [P in U]: never;
} & { [x: string]: never })[T];

export type Overwrite<T, U> = Pick<T, Diff<keyof T, keyof U>> & U;

export type ValuesOf<T> = T[keyof T];

//
export type UnionToIntersection<U> = (
  U extends any ? (k: U) => void : never
) extends (k: infer I) => void
  ? I
  : never;

export type Command = (props: CommandProps) => boolean;

export type CommandSpec = (...args: any[]) => Command;
export type AnyCommands = Record<string, (...args: any[]) => Command>;
export type KeysWithTypeOf<T, Type> = {
  [P in keyof T]: T[P] extends Type ? P : never;
}[keyof T];

export type UnionCommands<T = Command> = UnionToIntersection<
  ValuesOf<Pick<Commands<T>, KeysWithTypeOf<Commands<T>, {}>>>
>;

export type RawCommands = {
  [Item in keyof UnionCommands]: UnionCommands<Command>[Item];
};

export type SingleCommands = {
  [Item in keyof UnionCommands]: UnionCommands<boolean>[Item];
};

export type ChainedCommands = {
  [Item in keyof UnionCommands]: UnionCommands<ChainedCommands>[Item];
} & {
  run: () => boolean;
};

export type CanCommands = SingleCommands & { chain: () => ChainedCommands };

export type CommandProps = {
  editor: Editor;
  tr: Transaction;
  commands: SingleCommands;
  can: () => CanCommands;
  chain: () => ChainedCommands;
  state: EditorState;
  // view: EditorView;
  dispatch: ((args?: any) => any) | undefined;
};

export type ParentConfig<T> = Partial<{
  [P in keyof T]: Required<T>[P] extends (...args: any) => any
    ? (...args: Parameters<Required<T>[P]>) => ReturnType<Required<T>[P]>
    : T[P];
}>;

export type Attribute = {
  default: any;
  rendered?: boolean;
  renderHTML?: // | ((attributes: Record<string, any>) => Record<string, any> | null)
  null;
  parseHTML?: // ((element: HTMLElement) => any | null)
  null;
  keepOnSplit: boolean;
  isRequired?: boolean;
};

export type GlobalAttributes = {
  types: string[];
  attributes: {
    [key: string]: Attribute;
  };
}[];

export type AnyConfig = ExtensionConfig | NodeConfig | MarkConfig;
export type AnyExtension = Extension | Node | Mark;
export type Extensions = AnyExtension[];
export type EnableRules = (AnyExtension | string)[] | boolean;
export type ExtendedRegExpMatchArray = RegExpMatchArray & {
  data?: Record<string, any>;
};

export type Attributes = {
  [key: string]: Attribute;
};
export type Primitive =
  | null
  | undefined
  | string
  | number
  | boolean
  | symbol
  | bigint;

export type MaybeReturnType<T> = T extends (...args: any) => any
  ? ReturnType<T>
  : T;
export type MaybeThisParameterType<T> = Exclude<T, Primitive> extends (
  ...args: any
) => any
  ? ThisParameterType<Exclude<T, Primitive>>
  : any;
export type RemoveThis<T> = T extends (...args: any) => any
  ? (...args: Parameters<T>) => ReturnType<T>
  : T;

export type MarkRange = {
  mark: ProseMirrorMark;
  from: number;
  to: number;
};

export type NodeRange = {
  node: ProseMirrorNode;
  from: number;
  to: number;
};

export type ExtensionAttribute = {
  type: string;
  name: string;
  attribute: Required<Attribute>;
};
