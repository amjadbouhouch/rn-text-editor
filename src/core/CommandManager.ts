import {
  TextSelection,
  Transaction,
  type EditorState,
} from 'prosemirror-state';
import { editorHelper } from '../utils';
import type { Editor } from './Editor';
import type {
  AnyCommands,
  CanCommands,
  ChainedCommands,
  CommandProps,
  SingleCommands,
} from './types';

export class CommandManager {
  editor: Editor;
  // for now
  rawCommands: AnyCommands = {};

  customState?: EditorState;

  constructor(props: { editor: Editor; state?: EditorState }) {
    this.editor = props.editor;
    this.rawCommands = this.editor.extensionManager.commands;
    this.customState = props.state;
  }

  get hasCustomState(): boolean {
    return !!this.customState;
  }

  get state(): EditorState {
    return this.customState || this.editor.state;
  }

  get commands(): SingleCommands {
    const { rawCommands, state } = this;
    const { tr } = state;
    const props = this.buildProps(tr);

    return Object.fromEntries(
      Object.entries(rawCommands).map(([name, command]) => {
        const method = (...args: any[]) => {
          const callback = command(...args)(props);

          if (!tr.getMeta('preventDispatch') && !this.hasCustomState) {
            // view.dispatch(tr);
            this.dispatch(tr);
          }

          return callback;
        };

        return [name, method];
      })
    ) as unknown as SingleCommands;
  }

  get chain(): () => ChainedCommands {
    return () => this.createChain();
  }

  get can(): () => CanCommands {
    return () => this.createCan();
  }

  public createChain(
    startTr?: Transaction,
    shouldDispatch = true
  ): ChainedCommands {
    const { rawCommands, state } = this;
    // const { view } = editor;
    const callbacks: boolean[] = [];
    const hasStartTransaction = !!startTr;
    const tr = startTr || state.tr;

    const run = () => {
      if (
        !hasStartTransaction &&
        shouldDispatch &&
        !tr.getMeta('preventDispatch') &&
        !this.hasCustomState
      ) {
        this.dispatch(tr);
      }

      return callbacks.every((callback) => callback === true);
    };

    const chain = {
      ...Object.fromEntries(
        Object.entries(rawCommands).map(([name, command]) => {
          const chainedCommand = (...args: never[]) => {
            const props = this.buildProps(tr, shouldDispatch);
            const callback = command(...args)(props);

            callbacks.push(callback);

            return chain;
          };

          return [name, chainedCommand];
        })
      ),
      run,
    } as unknown as ChainedCommands;

    return chain;
  }

  public createCan(startTr?: Transaction): CanCommands {
    const { rawCommands, state } = this;
    const dispatch = false;
    const tr = startTr || state.tr;
    const props = this.buildProps(tr, dispatch);
    const formattedCommands = Object.fromEntries(
      Object.entries(rawCommands).map(([name, command]) => {
        return [
          name,
          (...args: never[]) =>
            command(...args)({ ...props, dispatch: undefined }),
        ];
      })
    ) as unknown as SingleCommands;

    return {
      ...formattedCommands,
      chain: () => this.createChain(tr, dispatch),
    } as CanCommands;
  }

  dispatch = (tr: Transaction) => {
    this.editor.dispatch(tr);
  };

  public buildProps(tr: Transaction, shouldDispatch = true): CommandProps {
    const { rawCommands, editor, state } = this;
    // const { view } = editor;

    const props: CommandProps = {
      tr,
      editor,
      // view,
      state: editorHelper.createChainableState({
        state,
        transaction: tr,
      }),
      dispatch: shouldDispatch ? () => undefined : undefined,
      chain: () => this.createChain(tr, shouldDispatch),
      can: () => this.createCan(tr),
      get commands() {
        return Object.fromEntries(
          Object.entries(rawCommands).map(([name, command]) => {
            return [name, (...args: never[]) => command(...args)(props)];
          })
        ) as unknown as SingleCommands;
      },
    };

    return props;
  }
  setSelection = (start: number, end?: number) => {
    const tr = this.editor.state.tr;
    const newStart = start + 1;
    const newEnd = end ? end + 1 : undefined;
    try {
      const newTransaction = tr.setSelection(
        TextSelection.create(this.state.doc, newStart, newEnd)
      );

      return this.dispatch(newTransaction);
    } catch (error) {
      console.error(error);
    }
  };

  handleKeyPress = (_: string) => {};
}
