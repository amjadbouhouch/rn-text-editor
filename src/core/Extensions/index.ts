export * from './commands';
export * from './extension-bold';
export * from './extension-italic';
