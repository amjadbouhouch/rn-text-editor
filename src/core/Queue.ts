type EventQueueProps<T> = {
  handleQueue: (payload: T) => Promise<void>;
};

export class Queue<T> {
  queue: T[] = [];
  isProcessing = false;
  handleQueue!: EventQueueProps<T>['handleQueue'];
  constructor({ handleQueue }: EventQueueProps<T>) {
    this.handleQueue = handleQueue;
    this.queue = [];
    this.isProcessing = false;
  }

  enqueue(payload: T) {
    this.queue.push(payload);
    if (!this.isProcessing) {
      this.processQueue();
    }
  }

  processQueue() {
    if (this.queue.length === 0) {
      this.isProcessing = false;
      return;
    }

    this.isProcessing = true;
    const currentEvent = this.queue.shift()!;
    this.handleQueue(currentEvent).then(() => {
      this.processQueue();
    });
  }
}
