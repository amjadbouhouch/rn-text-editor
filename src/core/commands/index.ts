export * from './clearNodes';
export * from './command';

declare module 'rn-text-editor' {
  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  export interface Commands<ReturnType = any> {}
}
