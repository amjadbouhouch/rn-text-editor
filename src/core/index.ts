export * from './CommandManager';
export * from './Editor';

//@ts-ignore
export interface ExtensionConfig<Options = any, Storage = any> {}

//@ts-ignore
export interface NodeConfig<Options = any, Storage = any> {}

//@ts-ignore
export interface MarkConfig<Options = any, Storage = any> {}
