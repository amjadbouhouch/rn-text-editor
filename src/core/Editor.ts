import { Fragment, Schema } from 'prosemirror-model';
import { marks } from 'prosemirror-schema-basic';
import { EditorState, Transaction } from 'prosemirror-state';
import type { EditorEvents, FocusPosition, JSONContent } from '../types';
import { editorHelper } from '../utils';
import { CommandManager } from './CommandManager';
import { EventEmitter } from './EventEmitter';

export type EditorProps = {
  initialContent: JSONContent[];
  focusPosition?: FocusPosition;
  onBeforeCreate: (props: EditorEvents['beforeCreate']) => void;
  onCreate: (props: EditorEvents['create']) => void;
  onUpdate: (props: EditorEvents['update']) => void;
  onSelectionUpdate: (props: EditorEvents['selectionUpdate']) => void;
  onTransaction: (props: EditorEvents['transaction']) => void;
  onFocus: (props: EditorEvents['focus']) => void;
  onBlur: (props: EditorEvents['blur']) => void;
  onDestroy: (props: EditorEvents['destroy']) => void;
};
export class Editor extends EventEmitter<EditorEvents> {
  public schema!: Schema;
  public commandManager!: CommandManager;
  public state!: EditorState;
  public options: EditorProps = {
    initialContent: [],
    focusPosition: 'end',
    onBeforeCreate: () => null,
    onCreate: () => null,
    onUpdate: () => null,
    onSelectionUpdate: () => null,
    onTransaction: () => null,
    onFocus: () => null,
    onBlur: () => null,
    onDestroy: () => null,
  };
  constructor(options: Partial<EditorProps>) {
    super();
    this.setOptions(options);
    //

    this.schema = new Schema({
      nodes: {
        doc: { content: 'block+' },
        paragraph: {
          content: 'inline*',
          group: 'block',
          attrs: { id: { default: null } },
          parseDOM: [
            {
              tag: 'p',
              // @ts-expect-error
              getAttrs: (dom) => ({ id: dom.getAttribute('id') }),
            },
          ],
          toDOM: (node) => ['p', { id: node.attrs.id }, 0],
        },
        text: { inline: true, group: 'inline' },
      },
      marks: marks,
    });
    const { initialContent, focusPosition } = this.options;
    const parsedContent = Fragment.fromArray(
      initialContent.map((item) => this.schema.nodeFromJSON(item))
    );
    const doc = this.schema.node('doc', null, parsedContent);

    this.on('beforeCreate', this.options.onBeforeCreate);
    this.emit('beforeCreate', { editor: this });
    this.on('create', this.options.onCreate);
    this.on('update', this.options.onUpdate);
    this.on('selectionUpdate', this.options.onSelectionUpdate);
    this.on('transaction', this.options.onTransaction);
    this.on('focus', this.options.onFocus);
    this.on('blur', this.options.onBlur);
    this.on('destroy', this.options.onDestroy);

    const selection = editorHelper.resolveFocusPosition(doc, focusPosition);
    this.state = EditorState.create({
      schema: this.schema,
      doc: doc,
      selection: selection ?? undefined,
    });
    this.commandManager = new CommandManager({ editor: this });
    //
    setTimeout(() => {
      this.emit('create', { editor: this });
    }, 0);
  }
  dispatch(tr: Transaction) {
    const newState = this.state.apply(tr);
    this.state = newState;
    this.emit('update', { editor: this, transaction: tr });
  }
  public setOptions(options: Partial<EditorProps>) {
    this.options = { ...this.options, ...options };
  }

  contentAsJson() {
    return this.state.doc.content.toJSON() as JSONContent[] | JSONContent;
  }
  getNativeText() {
    return this.state.doc.textContent;
  }
}
