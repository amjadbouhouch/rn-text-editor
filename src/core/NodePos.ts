import { Fragment, Node, ResolvedPos } from 'prosemirror-model';

import type { Content, Range } from './types';
import { Editor } from './Editor';

export class NodePos {
  private resolvedPos: ResolvedPos;

  private editor: Editor;

  constructor(pos: ResolvedPos, editor: Editor) {
    this.resolvedPos = pos;
    this.editor = editor;
  }

  get node(): Node {
    return this.resolvedPos.node();
  }

  get depth(): number {
    return this.resolvedPos.depth;
  }

  get pos(): number {
    return this.resolvedPos.pos;
  }

  get content(): Fragment {
    return this.node.content;
  }

  set content(__: Content) {
    // this.editor.commands.insertContentAt({ from: this.from, to: this.to }, content)
  }

  get attributes(): { [key: string]: any } {
    return this.node.attrs;
  }

  get textContent(): string {
    return this.node.textContent;
  }

  get size(): number {
    return this.node.nodeSize;
  }

  get from(): number {
    return this.resolvedPos.start(this.resolvedPos.depth);
  }

  get range(): Range {
    return {
      from: this.from,
      to: this.to,
    };
  }

  get to(): number {
    return (
      this.resolvedPos.end(this.resolvedPos.depth) + (this.node.isText ? 0 : 1)
    );
  }

  get parent(): NodePos | null {
    if (this.depth === 0) {
      return null;
    }

    const parentPos = this.resolvedPos.start(this.resolvedPos.depth - 1);
    const $pos = this.resolvedPos.doc.resolve(parentPos);

    return new NodePos($pos, this.editor);
  }

  get before(): NodePos | null {
    let $pos = this.resolvedPos.doc.resolve(this.from - 2);

    if ($pos.depth !== this.depth) {
      $pos = this.resolvedPos.doc.resolve(this.from - 3);
    }

    return new NodePos($pos, this.editor);
  }

  get after(): NodePos | null {
    let $pos = this.resolvedPos.doc.resolve(this.to + 2);

    if ($pos.depth !== this.depth) {
      $pos = this.resolvedPos.doc.resolve(this.to + 3);
    }

    return new NodePos($pos, this.editor);
  }

  get children(): NodePos[] {
    const children: NodePos[] = [];

    this.node.content.forEach((__, offset) => {
      const targetPos = this.pos + offset + 1;
      const $pos = this.resolvedPos.doc.resolve(targetPos);

      if ($pos.depth === this.depth) {
        return;
      }

      children.push(new NodePos($pos, this.editor));
    });

    return children;
  }

  get firstChild(): NodePos | null {
    return this.children[0] || null;
  }

  get lastChild(): NodePos | null {
    const children = this.children;

    return children[children.length - 1] || null;
  }

  closest(
    selector: string,
    attributes: { [key: string]: any } = {}
  ): NodePos | null {
    let node: NodePos | null = null;
    let currentNode = this.parent;

    while (currentNode && !node) {
      if (currentNode.node.type.name === selector) {
        if (Object.keys(attributes).length > 0) {
          const nodeAttributes = currentNode.node.attrs;
          const attrKeys = Object.keys(attributes);

          for (let index = 0; index < attrKeys.length; index += 1) {
            const key = attrKeys[index];

            if (key && nodeAttributes[key] !== attributes[key]) {
              break;
            }
          }
        } else {
          node = currentNode;
        }
      }

      currentNode = currentNode.parent;
    }

    return node;
  }

  querySelector(
    selector: string,
    attributes: { [key: string]: any } = {}
  ): NodePos | null {
    return this.querySelectorAll(selector, attributes, true)[0] || null;
  }

  querySelectorAll(
    selector: string,
    attributes: { [key: string]: any } = {},
    firstItemOnly = false
  ): NodePos[] {
    let nodes: NodePos[] = [];

    // iterate through children recursively finding all nodes which match the selector with the node name
    if (!this.children || this.children.length === 0) {
      return nodes;
    }

    this.children.forEach((node) => {
      if (node.node.type.name === selector) {
        if (Object.keys(attributes).length > 0) {
          const nodeAttributes = node.node.attrs;
          const attrKeys = Object.keys(attributes);

          for (let index = 0; index < attrKeys.length; index += 1) {
            const key = attrKeys[index];

            if (key && nodeAttributes[key] !== attributes[key]) {
              return;
            }
          }
        }

        nodes.push(node);

        if (firstItemOnly) {
          return;
        }
      }

      nodes = nodes.concat(node.querySelectorAll(selector));
    });

    return nodes;
  }

  setAttribute(_: { [key: string]: any }) {
    // const oldSelection = this.editor.state.selection;
    // this.editor.chain().setTextSelection(this.from).updateAttributes(this.node.type.name, attributes).setTextSelection(oldSelection.from)
    //   .run()
  }
}
