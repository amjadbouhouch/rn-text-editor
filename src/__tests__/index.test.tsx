import { Editor } from '../core/Editor';
let editor: Editor;

beforeEach(() => {
  editor = new Editor({
    initialContent: [
      {
        id: '1',
        type: 'paragraph',
        content: [
          {
            id: '1.1d',
            type: 'text',
            text: 'Hello',
          },
        ],
      },
    ],
  });
});

//
test('inset new text', () => {
  const newTr = editor.state.tr.insertText(' World 🎁');
  editor.commandManager.dispatch(newTr);
  expect(editor.state.doc.textContent).toBe('Hello World 🎁');
});

// test('Add new paragraph', () => {
//   const newTr = editor.state.tr;
//   const textNode = editor.state.schema.text('hello');
//   const paragraphNode = editor.state.schema.nodes.paragraph.create(
//     {
//       id: '2',
//     },
//     textNode,
//     null
//   );
//   newTr.insert(editor.state.doc.content.size, paragraphNode);
//   editor.commandManager.dispatch(newTr);

// });
