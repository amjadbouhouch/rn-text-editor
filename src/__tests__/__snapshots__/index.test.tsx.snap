// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "Hello",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`inject new text 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "Hello",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;
